subprojects/execution/src/main/java/org/gradle/internal/execution/DeferrableExecution.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.execution;
+
+import org.gradle.internal.Try;
+
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * An invocation that can potentially be deferred.
+ *
+ * @param <T> The type which will be computed.
+ */
+public interface DeferrableExecution<T> {
+
+    /**
+     * The result of the invocation when it is already available.
+     */
+    Optional<Try<T>> getCompleted();
+
+    /**
+     * Obtain the result of the invocation, either by returning the already known result or by computing it.
+     */
+    Try<T> get();
```

Comment:
💄 Maybe we could come up with more explicit naming. Currently `getCompleted` could actually seem to be the one that would await the computation before returning. Javadocs are there, but naming could be very helpful when in the moment.

This is especially pronounced when reading code that has `.get().get()` in it, and one of the innocent `get`'s can be the one that runs for a long time.

Just for reference, not sure if helpful here: https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-deferred/

Reply:
Addressed.